woocommerce/woocommerce-ios
# [Order form] Show order item data (not product data) where possible in order form

Closes: #11389
Closes: #11357
⚠️ Depends on #11442 ⚠️ 


## Why
When creating or editing an order, we should be using data from the `OrderItem` in the `Order` as much as possible, for a number of reasons:

* Consistency with the order detail screen, and with other platforms (Android, web)
* Clarity when the product price includes tax.
   * For example: Discounts are added to the price excluding tax, so showing the product price including tax creates confusion when adding discounts. The price from `OrderItem` excludes tax, making it clearer what amount the discount applies to.
* Clarity when a product bundle includes discounts on bundled items.
   * For example: A bundled item can be priced individually with a discount on the regular product price. Showing the regular product price makes it look more expensive than it is in the bundle, while the price from `OrderItem` shows the discounted bundle price.

The primary difference in the order form UI is the displayed price. Instead of showing the regular price from the product, you will see the order item price, which does _not_ include tax.

## How

* The view model for each row in an order (`CollapsibleProductRowCardViewModel`) is now initialized with the name, sku, and price provided in `OrderItem` instead of `Product` or `ProductVariation`. (See 537fc53122a66277c6bb0705d495a89a2ffc49a3)
* There are now convenience initializers for `CollapsibleProductRowCardViewModel` to create a view model from an order item and a product or variation, to simplify how it is created in `EditableOrderViewModel`.
* `OrderSynchronizer` and `ProductInputTransformer` are updated to include the input name, sku, and price, with fallbacks to the product name, sku, and price. This ensures those values are still shown while the order is syncing (otherwise they are blank).
   * Note: Product variation names are still blank while the order is syncing; that is consistent with current behavior and I couldn't find a straightforward way to resolve it here.
   * For some reason `price` is still showing the product price (rather than the order item price) while an order syncs. I didn't see a straightforward way to resolve this (to show the order item price when we have it), so for the sake of wrapping up this work, I redacted the price in the card view while the order is syncing remotely.

## Testing instructions


- Go to the Orders tab
- Tap on `+` to create an order
- Add at least one product to the order -> Confirm the product appears in the order form with the expected product details, and with price excluding tax (this will be different from the product price, if tax is included in the product price on your store)
- Expand any product card and confirm you can update the product as expected (change quantity, add discount, remove product)
- Tap "Create" to create the order -> Confirm the created order contains the products and quantities you added, and make a note of the prices shown in order details
- Tap "Edit" to edit the order -> Confirm the order loads with the expected products and quantities, and with prices matching what was shown in order details

## Screenshots

/|Before|After
-|-|-
Order Details (no change)|![Simulator Screenshot - iPhone 14 Pro - 2023-12-12 at 22 51 52](https://github.com/woocommerce/woocommerce-ios/assets/8658164/ee705e01-82c5-4802-8c66-7843f9541186)|![Simulator Screenshot - iPhone 14 Pro - 2023-12-12 at 22 40 39](https://github.com/woocommerce/woocommerce-ios/assets/8658164/39a44fab-4bc3-4bb2-b4dc-c0941869cdd6)
Products in order|![Simulator Screenshot - iPhone 14 Pro - 2023-12-12 at 22 51 56](https://github.com/woocommerce/woocommerce-ios/assets/8658164/32c2133b-ca88-4377-b683-0b395cdefc41)|![Simulator Screenshot - iPhone 14 Pro - 2023-12-12 at 22 40 43](https://github.com/woocommerce/woocommerce-ios/assets/8658164/8b78b6df-45d9-4796-b7dd-cef11dff6817)
Product with discount|![Simulator Screenshot - iPhone 14 Pro - 2023-12-12 at 22 52 02](https://github.com/woocommerce/woocommerce-ios/assets/8658164/54fa6671-9f3b-49da-ab3c-0ac53b47ba6c)|![Simulator Screenshot - iPhone 14 Pro - 2023-12-12 at 22 40 47](https://github.com/woocommerce/woocommerce-ios/assets/8658164/5926fa4d-230c-404b-8e50-fe38e181fb50)



https://github.com/woocommerce/woocommerce-ios/assets/8658164/ff38990c-acf3-45cf-a29d-42a365676c1e




---
